### DISCUSSION TITLE
Allow task names with wildcards in pipeline definition

### DISCUSSION THREAD

**Discussion by beeequeue** [vercel/turborepo#7496] [Ideas]
### Describe the feature you'd like to request

We have two main types of workspaces in our monorepo: packages and microfrontends. These two types each require different  `dependsOn` configurations, but from what I can see in the documentation this is not possible to do.

### Describe the solution you'd like

I propose being able to use the new `--filter` syntax in the Pipeline name configs:

```json
"./microfrontends/*#build": {
  "dependsOn": [
    "$APP_ENV",
    "src/**",
    "webpack.config.*"
  ],
  "outputs": ["dist/**"]
},
"./packages/*#build": {
  "dependsOn": [
	"$NODE_ENV",
    "src/**"
  ],
  "outputs": ["dist/**"]
},
"build": {
  "dependsOn": [
    "src/**"
  ],
  "outputs": ["dist/**"]
},
```

### Describe alternatives you've considered

It would be nice to at the very least have globs work in the names: `@mf/*#build`. Then we could name every workspace differently to handle these cases.

Another option would be to not require every string in `dependsOn` to match a file, then any missing files would be hashed as empty or something similar.

**Comment by silesky** [comment]
This may be related to a scenario I'm researching!

My workspaces are contained in `./examples/**` and `./packages/**`. I want to create a lint command that lints everything, but  _only_ dependsOn `./packages` (since examples depend on packages, but no example or package depends on any example).  AFAIK, I currently have to name every repo in `./packages` explicitly, like:
```
    "lint": {
      "dependsOn": [
        "package1#build",
        "package2#build",
        "package3#build",
        "package4#build",
      ],
      "outputs": [],
    }
```

I would like to do something like:
```
    "lint": {
      "dependsOn": [
        "./packages/**#build
      ],
      "outputs": [],
    }
```

**Comment by mehulkar** [comment]
This isn't as powerful as you want, but [Workspace Configs](https://turbo.build/repo/docs/core-concepts/monorepos/configuring-workspaces) can let you define different configs for the same task in different workspaces.

**Comment by mehulkar** [comment]
I've renamed this issue to more generally be about wildcarding, even though the original issue was specifically about using _directory paths_, rather than _package names_ in the task definitions.

**Comment by psychobolt** [comment]
Should we support single asterisk notation? If so, there could be a collision with: https://github.com/vercel/turbo/issues/4145

**Comment by mehulkar** [comment]
#4145's proposed solution may conflict, that's a good call out, but I think that issue is asking for a general "finalization" API. I think that could be a good feature as well, and we should take that into account when designing the API. 

The next step would be to outline the scope of this API and (as @psychobolt mentioned) where it could overlap with other features. Off the top of my head, `dependsOn` should be able to declare:

- More than one `task` name with a wildcard (e.g. `build:*`)
- More than one package name with a wildcard (e.g. `ui-*:build`)

It could overlap with:

- Finalization chains (e.g. `dependsOn: ["**everything**"]`) #4145
- `turbo run foo*` (https://github.com/vercel/turbo/issues/1029)
